data_steward/cdr_cleaner/clean_cdr_engine.py

```diff
@@ -140,6 +150,222 @@ def clean_dataset(project=None, statements=None, data_stage=stage.UNSPECIFIED):
+                ce_consts.QUERY_RUN_MESSAGE_TEMPLATE.render(
+                    index, query_count, **module_info_dict, **query_dict))
+            job_config = generate_job_config(client, query_dict)
+
+            module_name = module_info_dict[cdr_consts.MODULE_NAME]
+            query_job = client.query(query=query_dict.get(cdr_consts.QUERY),
+                                     job_config=job_config,
+                                     job_id_prefix=f'clean_{module_name}_')
+
+            # wait for job to complete
+            query_job.result()
+        except (GoogleCloudError, TOError) as exp:
+            LOGGER.exception(
+                ce_consts.FAILURE_MESSAGE_TEMPLATE.render(client.project,
+                                                          **module_info_dict,
+                                                          **query_dict,
+                                                          exception=exp))
+            failures.append(query_dict[cdr_consts.QUERY])
+            continue
+        else:
+            successes.append(query_dict[cdr_consts.QUERY])
+    return successes, failures
+
+
+def get_module_info(query_function):
+    """
+    Extracts module information about the cleaning rule
+
+    :param query_function: a function that generates a list of query dictionaries
+    :return: dictionary containing the module information
+        keys: module_name, function_name, line_no
+    """
+    function_name = query_function.__name__
+    module_name = inspect.getmodule(query_function).__name__
+    _, line_no = inspect.getsourcelines(query_function)
+    module_info_dict = {
+        cdr_consts.MODULE_NAME: module_name,
+        cdr_consts.FUNCTION_NAME: function_name,
+        cdr_consts.LINE_NO: line_no
+    }
+    return module_info_dict
+
+
+def infer_rule(clazz, project_id, dataset_id, sandbox_dataset_id, client=None):
+    """
+    Identify the type of the clean rule set up if client is provided
+
+    :param clazz: Clean rule class or old style clean function
+    :param project_id: identifies the project
+    :param dataset_id: identifies the dataset to clean
+    :param sandbox_dataset_id: identifies the dataset to store sandbox tables
+    :param client: BQ client
+    :return: query_list: list of query_dicts
+             query_function: function that generated query_list
+    """
+    if issubclass(clazz, BaseCleaningRule):
+        instance = clazz(project_id, dataset_id, sandbox_dataset_id)
+
+        query_function = instance.get_query_specs
+
+        if client is not None:
+            instance.setup_rule(client)
+
+        query_list = instance.get_query_specs()
+    else:
+        query_function = clazz
+        try:
+            query_list = clazz(project_id, dataset_id, sandbox_dataset_id)
+        except TypeError:
+            # raised when called with the 3 parameters and only 2 are needed
+            query_list = clazz(project_id, dataset_id)
+    return query_list, query_function
```

Comment:
I think the expectation is that this [well-named] function `infer_rule()` would be side-effect free. I recommend removing the client parameter and deferring invocation of `setup_rule()` to whoever is calling this func. Right now It's a bit awkward that this calls the query list func while also returning it for someone else to use- of course these are remnants of the past. Another consequence is that `setup_rule()` calls are not set to occur in the sequence we discussed (e.g. `rule_1.setup(), rule_1.query1(), rule_1.query2(), rule_2.setup() ...`).

What do you think of moving the call to `get_module_info()` here? It would seem appropriate to me at the moment as the purpose of this function seems to be to assemble some universal representation of old and new style rules. Also right now the "clean_dataset" use case is getting the module info (just for logging) but the "get_rules_documentation" use case (where this would shine brightest) does not currently get module info. 

To solve this, I'm wondering if we could figure out a way for this function to return a data structure similar to this:
```
(query_spec_func, setup_rule_func, module_info)
```

For new-style rules it be relatively straightforward:
```
return instance.get_query_specs, instance.setup_rule, get_module_info(instance.get_query_specs)
```
Old-style rules might need a little creativity. Something like this may work:
```
def old_style_query_specs():
  try:
    query_list = clazz(project_id, dataset_id, sandbox_dataset_id)
  except TypeError:
    # raised when called with the 3 parameters and only 2 are needed
    query_list = clazz(project_id, dataset_id)
  return query_list
def empty_setup_rule(client):
  pass

return old_style_query_specs, empty_setup_rule, get_module_info(instance.get_query_specs)
```
and `project_id`, `dataset_id`, `sandbox_dataset_id` should evaluate fine if I understand Python closures correctly. But even if not there's different ways to encapsulate them in another closure. This comment is heavy, let's discuss it soon.

Reply:
Possible test harness on closure:
```
project_id = 'fake_project'
dataset_id = 'fake_dataset'
sandbox_dataset_id = 'fake_sandbox'
query_func, setup_rule, module_info = infer_rule(clazz, project_id, dataset_id, sandbox_dataset_id)
project_id = 'another_project_id'
query_func()
```

Reply:
Thanks for the detailed explanation, Mark! I have created the closure, please let me know if anything looks off. I will create the test soon.